test/server/connection_handler_test.cc

```diff
@@ -746,6 +747,148 @@ TEST_F(ConnectionHandlerTest, FallbackToWildcardListener) {
+  Network::Address::InstanceConstSharedPtr normal_address(
+      new Network::Address::Ipv4Instance("127.0.0.1", 10001));
+  EXPECT_CALL(*socket_factory_, localAddress()).WillRepeatedly(ReturnRef(normal_address));
+  handler_->addListener(absl::nullopt, *test_listener1);
+
+  Network::TcpListenerCallbacks* ipv4_any_listener_callbacks;
+  auto listener2 = new NiceMock<Network::MockListener>();
+  TestListener* ipv4_any_listener = addListener(1, false, false, "ipv4_any_test_listener",
+                                                listener2, &ipv4_any_listener_callbacks);
+  Network::Address::InstanceConstSharedPtr any_address(
+      new Network::Address::Ipv4Instance("0.0.0.0", 80));
+  EXPECT_CALL(*socket_factory_, localAddress()).WillRepeatedly(ReturnRef(any_address));
+  handler_->addListener(absl::nullopt, *ipv4_any_listener);
+
+  Network::TcpListenerCallbacks* ipv6_any_listener_callbacks;
+  auto listener3 = new NiceMock<Network::MockListener>();
+  TestListener* ipv6_any_listener = addListener(1, false, false, "ipv6_any_test_listener",
+                                                listener3, &ipv6_any_listener_callbacks);
+  Network::Address::InstanceConstSharedPtr any_address_ipv6(
+      new Network::Address::Ipv6Instance("::", 80));
+  EXPECT_CALL(*socket_factory_, localAddress()).WillRepeatedly(ReturnRef(any_address_ipv6));
+  handler_->addListener(absl::nullopt, *ipv6_any_listener);
+
+  Network::MockListenerFilter* test_filter = new Network::MockListenerFilter();
+  EXPECT_CALL(*test_filter, destroy_());
+  Network::MockConnectionSocket* accepted_socket = new NiceMock<Network::MockConnectionSocket>();
+  bool redirected = false;
+  EXPECT_CALL(factory_, createListenerFilterChain(_))
+      .WillRepeatedly(Invoke([&](Network::ListenerFilterManager& manager) -> bool {
+        // Insert the Mock filter.
+        if (!redirected) {
+          manager.addAcceptFilter(listener_filter_matcher_,
+                                  Network::ListenerFilterPtr{test_filter});
+          redirected = true;
+        }
+        return true;
+      }));
+
+  Network::Address::InstanceConstSharedPtr alt_address(
+      new Network::Address::Ipv6Instance("::2", 80, nullptr));
+  EXPECT_CALL(*test_filter, onAccept(_))
+      .WillOnce(Invoke([&](Network::ListenerFilterCallbacks& cb) -> Network::FilterStatus {
+        cb.socket().addressProvider().restoreLocalAddress(alt_address);
+        return Network::FilterStatus::Continue;
+      }));
+  EXPECT_CALL(manager_, findFilterChain(_)).WillOnce(Return(filter_chain_.get()));
+  auto* connection = new NiceMock<Network::MockServerConnection>();
+  EXPECT_CALL(dispatcher_, createServerConnection_()).WillOnce(Return(connection));
+  EXPECT_CALL(factory_, createNetworkFilterChain(_, _)).WillOnce(Return(true));
+  listener_callbacks1->onAccept(Network::ConnectionSocketPtr{accepted_socket});
+  EXPECT_EQ(1UL, handler_->numConnections());
+
+  // Add negative test case check
```

Comment:
cc @lambdai There is where I'm struggling figuring out how to make sure the right listener is being used.

I had added some logging locally and was able to verify the correct listener was being returned, but a stronger check is needed. I was able to track it down to https://github.com/envoyproxy/envoy/blob/d526e38a79d643710d2236322abf7c0804e4146d/source/server/active_tcp_listener.cc#L239 as probably the best way to make sure the correct listener is needed with an `EXPECT_CALL`.

Something similar to
```
EXPECT_CALL(balancer1, pickTargetHandler(_))
      .WillOnce(testing::DoAll(
          testing::WithArg<0>(Invoke([](auto& target) { target.incNumConnections(); })),
          ReturnRef(*active_listener1)));
```

But some of the difficulties are the `connection_handler_test.cc` and `active_tcp_listener_test.cc` both approach similar testing differently. I could move this test into `active_tcp_listener_test.cc` if you'd think that'd be sufficient. I'd prefer to keep it close the `ConnectionHandlerImpl`, but at this point I'm open to move it. Thanks!

Reply:
IMO The old behavior is not that critical.